lib/c-api/src/wasm_c_api/externals/mod.rs

```diff
@@ -6,16 +6,150 @@ mod table;
 pub use function::*;
 pub use global::*;
 pub use memory::*;
-use std::sync::Arc;
+use std::mem;
 pub use table::*;
-use wasmer::{Extern, Instance};
+use wasmer::{Extern, ExternType};
 
 #[allow(non_camel_case_types)]
-#[derive(Clone)]
+#[repr(transparent)]
 pub struct wasm_extern_t {
-    // this is how we ensure the instance stays alive
-    pub(crate) instance: Option<Arc<Instance>>,
-    pub(crate) inner: Extern,
+    pub(crate) inner: wasm_extern_inner,
+}
+
+/// All elements in this union must be `repr(C)` and have a
+/// `CApiExternTag` as their first element.
+#[allow(non_camel_case_types)]
+pub(crate) union wasm_extern_inner {
+    function: mem::ManuallyDrop<wasm_func_t>,
+    memory: mem::ManuallyDrop<wasm_memory_t>,
+    global: mem::ManuallyDrop<wasm_global_t>,
+    table: mem::ManuallyDrop<wasm_table_t>,
+}
+
+#[cfg(test)]
+mod extern_tests {
+    use super::*;
+
+    #[test]
+    fn externs_are_the_same_size() {
+        use std::mem::size_of;
+        assert_eq!(size_of::<wasm_extern_t>(), size_of::<wasm_func_t>());
+        assert_eq!(size_of::<wasm_extern_t>(), size_of::<wasm_memory_t>());
+        assert_eq!(size_of::<wasm_extern_t>(), size_of::<wasm_global_t>());
+        assert_eq!(size_of::<wasm_extern_t>(), size_of::<wasm_table_t>());
+    }
+}
+
+impl Drop for wasm_extern_inner {
+    fn drop(&mut self) {
+        unsafe {
+            let tag = self.function.tag;
+            dbg!(tag);
+            match tag {
+                CApiExternTag::Function => mem::ManuallyDrop::drop(dbg!(&mut self.function)),
+                CApiExternTag::Global => mem::ManuallyDrop::drop(dbg!(&mut self.global)),
+                CApiExternTag::Table => mem::ManuallyDrop::drop(&mut self.table),
+                CApiExternTag::Memory => mem::ManuallyDrop::drop(dbg!(&mut self.memory)),
```

Comment:
Be careful with the `dbg!` here:

```suggestion
                CApiExternTag::Function => mem::ManuallyDrop::drop(&mut self.function),
                CApiExternTag::Global => mem::ManuallyDrop::drop(&mut self.global),
                CApiExternTag::Table => mem::ManuallyDrop::drop(&mut self.table),
                CApiExternTag::Memory => mem::ManuallyDrop::drop(&mut self.memory),
```